File: src/main/java/sample/design_pattern/ChainOfResponsibility.java
/*
 * Copyright 2013 SatoShun
 *
 * Chain of responsibility Pattern
 */

package sample.design_pattern;


import java.util.*;


interface IEventHandler {
    boolean process(String message);
}


class Boss implements IEventHandler {
    public boolean process(String message) {
        if (message.equals("boss")) {
            System.out.println("boss appears");
            return true;
        }
        return false;
    }
}


class Enemy implements IEventHandler {
    public boolean process(String message) {
        if (message.contains("enemy")) {
            System.out.println("enemy appears");
            return true;
        }

        return false;
    }
}


class EnemyController {
    List<IEventHandler> handlerList;

    public EnemyController() {
        handlerList = new ArrayList<IEventHandler>();
    }

    public void setEnemy(IEventHandler handler) {
        handlerList.add(handler);
    }

    public void sendMessage(String message) {
        for (IEventHandler handler : handlerList) {
            if (handler.process(message)) {
                break;
            }
        }
    }
}


class ChainOfResponsibility {
    public static void main(String[] args) {
        EnemyController controller = new EnemyController();

        controller.setEnemy(new Enemy());
        controller.setEnemy(new Boss());

        controller.sendMessage("enemy");
        controller.sendMessage("boss");
    }
}

File: src/main/scala/sample/design_pattern/ChainOfResponsibility.scala
/*
 * Copyright 2013 SatoShun
 *
 * Chain of responsibility Pattern
 */

package sample.design_pattern;


trait IEventHandler {
    def process(key: String, message: String);
}


class HelpMessageBox extends IEventHandler {
    private var child: IEventHandler = null

    def addChild(child: IEventHandler) = this.child = child

    def process(key: String, message: String) = {
        if (key == "Help") {
            println("Operate HelpMessageBox " + message)
        } else {
            child.process(key, message)
        }
    }
}


class DefaultMessageBox extends IEventHandler {
    def process(key: String, message: String) = {
        println("Operate Default " + message)
    }
}


class MessageController(val messageBox: IEventHandler) {
    def sendMessage(key: String, message: String) = {
        messageBox.process(key, message)
    }
}


object ChainOfResponsibility {
    def main(args: Array[String]) = {
        val messageBox = new HelpMessageBox();
        messageBox.addChild(new DefaultMessageBox())
        val messageController = new MessageController(messageBox)
        messageController.sendMessage("Help", "please help me")
        messageController.sendMessage("Hoge", "hogehoge")
    }
}
